train/compute/python/tools/trace_link.py

```diff
@@ -0,0 +1,309 @@
+                exit(0)
+    else:
+        print('Ops count mismatch between kineto and execution trace')
+
+    return et_enhanced
+
+
+def approximate_match(kineto_et_events, et_nodes):
+    # Since kineto trace is missing the annotations for processes/threads, we add them back to match with ET
+    kineto_event_per_thread = {}
+
+    # Mapping node id to the corresponding node
+    kineto_nodes_mapping = {}
+
+    for event in kineto_et_events:
+        if event['tid'] not in kineto_event_per_thread:
+            kineto_event_per_thread[event['tid']] = []
+        kineto_event_per_thread[event['tid']].append(event)
+    
+    start_time = kineto_et_events[0]['ts']
+    end_time = -1
+    for event in kineto_et_events:
+        end_time = max(end_time, event['ts'] + event['dur'])
+    process_node = Kineto_node('[pytorch|profiler|execution_graph|process]', start_time, end_time, 0)
+    kineto_nodes_mapping[0] = process_node
+
+    cnt = 1
+    for thread in kineto_event_per_thread:
+        start_time = kineto_event_per_thread[thread][0]['ts']
+        end_time = -1
+        for event in kineto_event_per_thread[thread]:
+            end_time = max(end_time, event['ts'] + event['dur'])
+
+        thread_node = Kineto_node('[pytorch|profiler|execution_graph|thread]', start_time, end_time, cnt)
+        kineto_nodes_mapping[cnt] = thread_node
+        cnt += 1
+
+        process_node.children.append(thread_node)
+
+        kineto_nodes = [thread_node]
+        for event in kineto_event_per_thread[thread]:
+            if event['ts'] < kineto_nodes[-1].end:
+                tmp = Kineto_node(event['name'], event['ts'], event['ts'] + event['dur'], cnt)
+                kineto_nodes_mapping[cnt] = tmp
+                cnt += 1
+                kineto_nodes[-1].children.append(tmp)
+                kineto_nodes.append(tmp)
+            else:
+                while kineto_nodes[-1].end <= event['ts']:
+                    kineto_nodes.pop()
+                tmp = Kineto_node(event['name'], event['ts'], event['ts'] + event['dur'], cnt)
+                kineto_nodes_mapping[cnt] = tmp
+                cnt += 1
+                kineto_nodes[-1].children.append(tmp)
+                kineto_nodes.append(tmp)
+
+    # Max call stack depth when building the tree, the deeper the more accurate but takes longer time
+    depth = 10
+
+    # Build a tree from the kineto trace
+    kineto_graph = transform_to_graph_depth(process_node, depth)
+    print("Kineto tree nodes number: ", len(kineto_graph.nodes))
+
+    # Build a tree from the execution trace
+    et_graph = transform_to_graph_depth(et_nodes[0], depth)
+    print("ET tree nodes number: ", len(et_graph.nodes))
+
+    # print("Kineto nodes info: ", list(kineto_graph.nodes(data=True)))
+    # print("ET nodes info: ", list(et_graph.nodes(data=True)))
+
+    # Create the GraphMatcher
+    GM = isomorphism.GraphMatcher(kineto_graph, et_graph)
+
+    # Duration of ET nodes
+    et_enhanced = {}
```

Comment:
since we are not returning nodes but a mapping of et node -> duration we could rename this field to 'et_enhanced_duration' ?